File: packages/contracts-python/src/aegis_contracts/parsing.py
"""Helpers for parsing contracts with structured errors."""

from __future__ import annotations

from pydantic import BaseModel, ValidationError

from aegis_contracts.errors import ContractErrorCode, ContractValidationError


def parse_contract[T: BaseModel](model: type[T], data: object) -> T:
    try:
        return model.model_validate(data)
    except ContractValidationError:
        raise
    except ValidationError as exc:
        raise ContractValidationError(
            code=ContractErrorCode.VALIDATION_FAILED,
            message="Contract validation failed",
            details={"errors": exc.errors()},
        ) from exc

File: packages/contracts-ts/src/parsing.ts
import { z } from 'zod';

import { ContractErrorCode, ContractValidationError } from './errors';

export function parseContract<T extends z.ZodTypeAny>(schema: T, data: unknown): z.infer<T> {
  try {
    // Zod ties parse output to the caller-provided schema generic.
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return -- schema.parse is parameterized by T
    return schema.parse(data);
  } catch (error) {
    if (error instanceof z.ZodError) {
      throw new ContractValidationError({
        code: ContractErrorCode.VALIDATION_FAILED,
        message: 'Contract validation failed',
        details: { errors: error.issues },
      });
    }
    throw error;
  }
}

export function safeParseContract<T extends z.ZodTypeAny>(
  schema: T,
  data: unknown,
): z.SafeParseReturnType<unknown, z.infer<T>> {
  return schema.safeParse(data);
}
